### DISCUSSION TITLE
Better solution on tensor.unsqueeze function available

### DISCUSSION THREAD

**Discussion by robberpen** [nnstreamer/nnstreamer#4608] [Q&A] [answered]
Hi Team

Based on https://github.com/nnstreamer/nnstreamer/issues/3777

My voice model input as tensor **float32[1,16000,1]**, but element alsasrc is **[1,16000]**
Is any tensor **unsqueeze** feature in nnstreamer? Or what is better solution such as customize unsqueeze feature on tensor_filter/tensor_transform element?

**Comment by anyj0527** [comment]
It the pipeline does not work, you may try adding `other/tensors,format=static,num_tensors=1,types=float32,dimensions=1:16000:1` before `tensor_filter`.

**Comment by myungjoo** [comment, marked as answer]
You don't need to unsqueeze it. ```N:M``` is dealt equivalent to ```N:M:1```. However, if  you want to unsqueeze ```1:N:M``` to ```N:M```, use ```tensor_transform```.

**Reply by robberpen** [reply]
Thanks your replay, and yes, it don't need to unsqueeze it, and working in my case now.
However, an error occurred on if I connect to tensor_decoder mode=image_labeling which it accept 2dim.
I just modified tensordec-imagelabel.c below and working now, No idea any side effect.

tensordec-imagelabel.c
```
il_setOption()
   g_return_val_if_fail (dim[0] > 0 && dim[1] == 1, NULL);
-  for (i = 2; i < NNS_TENSOR_RANK_LIMIT; i++)
+  for (i = 3; i < NNS_TENSOR_RANK_LIMIT; i++)
     g_return_val_if_fail (dim[i] == 0, NULL);
```

**Reply by myungjoo** [reply]
It appears that it needs to be updated like this:
```
-  g_return_val_if_fail (dim[i] == 0, NULL);
+  g_return_val_if_fail (dim[i] == 0 || dim[i] == 1, NULL);
```
@niley7464 Please take a look at this.
